src/brain/index_selection.cpp

```diff
@@ -0,0 +1,449 @@
+                                           IndexConfiguration &top_indexes,
+                                           Workload &workload) {
+  // Get the best m index configurations using the naive enumeration algorithm
+  // The naive algorithm gets all the possible subsets of size <= m and then
+  // returns the cheapest m indexes
+  assert(context_.naive_enumeration_threshold_ <= indexes.GetIndexCount());
+
+  // Define a set ordering of (index config, cost) and define the ordering in
+  // the set
+  std::set<std::pair<IndexConfiguration, double>, IndexConfigComparator>
+      running_index_config(workload);
+  std::set<std::pair<IndexConfiguration, double>, IndexConfigComparator>
+      temp_index_config(workload);
+  std::set<std::pair<IndexConfiguration, double>, IndexConfigComparator>
+      result_index_config(workload);
+  IndexConfiguration new_element;
+
+  // Add an empty configuration as initialization
+  IndexConfiguration empty;
+  // The running index configuration contains the possible subsets generated so
+  // far. It is updated after every iteration
+  running_index_config.insert({empty, 0.0});
+
+  for (auto index : indexes.GetIndexes()) {
+    // Make a copy of the running index configuration and add each element to it
+    temp_index_config = running_index_config;
+
+    for (auto t : temp_index_config) {
+      new_element = t.first;
+      new_element.AddIndexObject(index);
+
+      // If the size of the subset reaches our threshold, add to result set
+      // instead of adding to the running list
+      if (new_element.GetIndexCount() >=
+          context_.naive_enumeration_threshold_) {
+        result_index_config.insert(
+            {new_element, ComputeCost(new_element, workload)});
+      } else {
+        running_index_config.insert(
+            {new_element, ComputeCost(new_element, workload)});
+      }
+    }
+  }
+
+  // Put all the subsets in the result set
+  result_index_config.insert(running_index_config.begin(),
+                             running_index_config.end());
+  // Remove the starting empty set that we added
+  result_index_config.erase({empty, 0.0});
+
+  // Since the insertion into the sets ensures the order of cost, get the first
+  // m configurations
+  for (auto index_pair : result_index_config) {
+    top_indexes.Merge(index_pair.first);
+  }
+}
+
+void IndexSelection::GetAdmissibleIndexes(parser::SQLStatement *query,
+                                          IndexConfiguration &indexes) {
+  // Find out the indexable columns of the given workload.
+  // The following rules define what indexable columns are:
+  // 1. A column that appears in the WHERE clause with format
+  //    ==> Column OP Expr <==
+  //    OP such as {=, <, >, <=, >=, LIKE, etc.}
+  //    Column is a table column name.
+  // 2. GROUP BY (if present)
+  // 3. ORDER BY (if present)
+  // 4. all updated columns for UPDATE query.
+
+  union {
```

Comment:
Using the union here is a bit weird since you can just declare local variables in each of the switch cases. The compiler should be able to recognize that life time don't overlap and do this automatically. Using a somewhat esoteric feature like the union tends to confuse people.

Reply:
Fixed